File: ios/Plugin/Response.swift
//
//  Response.swift
//  Plugin
//
//  Created by Polo Swelsen on 28/03/2023.
//  Copyright © 2023 ilionx B.V. All rights reserved.
//

import Foundation
import Onfido

// 📝 Protocols are for testing purposes since SDK types are final
protocol CapacitorDocumentSideResult {
    var id: String { get }
}

protocol CapacitorDocumentResult {
    var capacitorFront: CapacitorDocumentSideResult { get }
    var capacitorBack: CapacitorDocumentSideResult? { get }
    var capacitorNfcMediaId: String? { get }
}

protocol CapacitorFaceResult {
    var id: String { get }
}

extension DocumentSideResult: CapacitorDocumentSideResult {}
extension DocumentResult: CapacitorDocumentResult {
    var capacitorFront: CapacitorDocumentSideResult { front }
    var capacitorBack: CapacitorDocumentSideResult? { back }
    var capacitorNfcMediaId: String? { nfcMediaId }
}

extension FaceResult: CapacitorFaceResult {}

func createResponse(_ results: [OnfidoResult], faceVariant: String?) -> [String: [String: Any]] {
    let document: DocumentResult? = results.compactMap { result in
        guard case let .document(documentResult) = result else { return nil }
        return documentResult
    }.first
    let face: FaceResult? = results.compactMap { result in
        guard case let .face(faceResult) = result else { return nil }
        return faceResult
    }.first
    return createResponse(document: document, face: face, faceVariant: faceVariant)
}

func createResponse(
    document: CapacitorDocumentResult? = nil,
    face: CapacitorFaceResult? = nil,
    faceVariant: String? = nil
) -> [String: [String: Any]] {
    var response = [String: [String: Any]]()

    if let documentResponse = document {
        response["document"] = ["front": ["id": documentResponse.capacitorFront.id]]
        if let backId = documentResponse.capacitorBack?.id,
           backId != documentResponse.capacitorFront.id
        {
            response["document"]?["back"] = ["id": documentResponse.capacitorBack?.id]
        }
        
        if let nfcId = documentResponse.capacitorNfcMediaId
        {
            response["document"]?["nfcMediaId"] = ["id": nfcId]
        }
    }

    if let faceResponse = face {
        var faceResponse = ["id": faceResponse.id]

        if let faceVariant = faceVariant {
            faceResponse["variant"] = faceVariant
        }

        response["face"] = faceResponse
    }

    return response
}

File: android/src/main/java/com/viveapp/onfido/capacitor/Response.kt
package com.viveapp.onfido.capacitor

import com.getcapacitor.JSObject

public class Response(
    frontId: String?,
    backId: String?,
    nfcMediaId: String?,
    faceId: String?,
    faceVariant: String?
) {
    private var document: Document?
    private var face: Face?

    init {
        document = initDocument(frontId, backId, nfcMediaId)
        face = initFace(faceId, faceVariant)
    }

    private fun initDocument(frontId: String?, backId: String?, nfcMediaId: String?): Document? {
        if (frontId != null || backId != null || nfcMediaId != null) {
            val document = Document()
            frontId?.let {
                document.front = Identifiable(it)
            }
            backId?.let {
                document.back = Identifiable(it)
            }
            nfcMediaId?.let {
                document.nfcMediaId = Identifiable(it)
            }
            return document
        }
        return null
    }

    private fun initFace(faceId: String?, faceVariant: String?): Face? {
        if (faceId != null && faceVariant != null) {
            return Face(faceId, faceVariant)
        }
        return null
    }

    fun toJSObject(): JSObject {
        val obj = JSObject()
        document?.let {
            if (it.front != null || it.back != null || it.nfcMediaId != null) {
                val doc = JSObject()
                it.front?.let { frontId ->
                    doc.put("front", JSObject().put("id", frontId))
                }
                it.back?.let { backId ->
                    doc.put("back", JSObject().put("id", backId))
                }
                it.nfcMediaId?.let { nfcMediaId ->
                    doc.put("nfcMediaId", JSObject().put("id", nfcMediaId))
                }
                obj.put("document", doc)
            }
        }
        face?.let {
            val face = JSObject()
            face.put("id", it.id)
            face.put("variant", it.variant)
            obj.put("face", face)
        }
        return obj
    }
}

open class Identifiable(var id: String = "default")
class Document {
    var front: Identifiable? = null
    var back: Identifiable? = null
    var nfcMediaId: Identifiable? = null
}

class Face(id: String, var variant: String) : Identifiable(id)